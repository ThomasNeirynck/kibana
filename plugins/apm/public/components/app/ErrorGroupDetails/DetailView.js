import React from 'react';
import styled from 'styled-components';
import {
  unit,
  units,
  px,
  colors,
  fontSizes,
  borderRadius
} from '../../../style/variables';
import { get } from 'lodash';
import { STATUS } from '../../../constants';
import {
  PropertiesTable,
  getLevelOneProps
} from '../../shared/PropertiesTable';
import { RelativeLink } from '../../../utils/url';

import moment from 'moment';
import Tab from '../../shared/Tab';
import Stacktrace from './Stacktrace';

const Container = styled.div`
  position: relative;
  border: 1px solid ${colors.elementBorder};
  border-radius: ${borderRadius};
  padding: ${px(units.plus)};
`;

const Header = styled.div`
  display: flex;
  justify-content: space-between;
`;

const Properties = styled.div`
  display: flex;
  margin: ${px(units.double)} 0;
  width: 100%;
  justify-content: space-between;
  flex-flow: row wrap;
`;

const Property = styled.div`
  width: 30%;
  margin-bottom: ${px(units.plus)};
`;

const PropertyLabel = styled.div`margin-bottom: ${px(units.quarter)};`;

const PropertyValue = styled.div`
  font-weight: bold;
  font-size: ${fontSizes.large};
`;

function AllOccurrencesLink({ errorGroup, appName }) {
  const groupId = errorGroup.group_id;
  const occurrencesCount = errorGroup.occurrences_count;

  if (!appName || !groupId) {
    return null;
  }

  const DiscoverLink = styled.a`margin: ${px(units.plus)} ${px(unit)} 0;`;

  const DiscoverIcon = styled.img`
    height: 18px;
    margin-top: -3px;
    margin-right: 6px;
    background: #005571;
    padding: 2px;
    border-radius: 4px;
    vertical-align: middle;
  `;

  const discoverRoute = `/app/kibana#/discover?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:now-24h,mode:quick,to:now))&_a=(interval:auto,query:(language:lucene,query:'context.app.name:${appName}%20AND%20error.grouping_key:${groupId}'),sort:!('@timestamp',desc))`;
  return (
    <DiscoverLink href={discoverRoute}>
      <DiscoverIcon src="/plugins/kibana/assets/discover.svg" />
      View {occurrencesCount} occurrences
    </DiscoverLink>
  );
}

const DEFAULT_TAB = 'stacktrace';

// Ensure the selected tab exists or use the default
function getCurrentTab(tabs = [], detailTab) {
  return tabs.includes(detailTab) ? detailTab : DEFAULT_TAB;
}

function getTabs(errorGroup) {
  const dynamicProps = Object.keys(errorGroup.data.error.context);
  return getLevelOneProps(dynamicProps);
}

function DetailView({ errorGroup, urlParams }) {
  if (errorGroup.status !== STATUS.SUCCESS) {
    return null;
  }

  const { appName } = urlParams;
  const timestamp = moment(get(errorGroup, 'data.error.@timestamp')).format();

  const tabs = getTabs(errorGroup);
  const currentTab = getCurrentTab(tabs, urlParams.detailTab);

  return (
    <Container>
      <Header>
        <h3>Error occurrence</h3>
        <AllOccurrencesLink errorGroup={errorGroup.data} appName={appName} />
      </Header>
      <Properties>
        <Property>
          <PropertyLabel>Occurrence recorded</PropertyLabel>
          <PropertyValue>{timestamp}</PropertyValue>
        </Property>
        {[1, 2, 3, 4, 5].map(item => {
          return (
            <Property key={item}>
              <PropertyLabel>Data label</PropertyLabel>
              <PropertyValue>Value</PropertyValue>
            </Property>
          );
        })}
      </Properties>
      {[DEFAULT_TAB, ...tabs].map(key => {
        return (
          <Tab selected={currentTab === key} key={key}>
            <RelativeLink query={{ detailTab: key }}>
              {key.toUpperCase()}
            </RelativeLink>
          </Tab>
        );
      })}

      <div>
        {currentTab === DEFAULT_TAB ? (
          <Stacktrace
            stacktraces={errorGroup.data.error.error.exception.stacktrace}
          />
        ) : (
          <PropertiesTable
            propData={errorGroup.data.error.context[currentTab]}
            propKey={currentTab}
          />
        )}
      </div>
    </Container>
  );
}

export default DetailView;
