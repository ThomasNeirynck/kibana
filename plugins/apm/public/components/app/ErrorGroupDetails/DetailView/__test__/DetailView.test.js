import React from 'react';
import toDiffableHtml from 'diffable-html';
import { mount } from 'enzyme';
import DetailView from '../index';
import props from './props.json';
import {
  mountWithRouterAndStore,
  mockMoment
} from '../../../../../utils/testHelpers';

describe('DetailView', () => {
  let storeState;
  beforeEach(() => {
    // Avoid timezone issues
    mockMoment();

    storeState = {
      location: { search: '' }
    };
  });

  it('should render empty state', () => {
    const wrapper = mount(
      <DetailView errorGroup={[]} urlParams={props.urlParams} />,
      storeState
    );

    expect(toDiffableHtml(wrapper.html())).toMatchSnapshot();
  });

  it('should render with data', () => {
    const wrapper = mountWithRouterAndStore(
      <DetailView errorGroup={props.errorGroup} urlParams={props.urlParams} />,
      storeState
    );

    expect(toDiffableHtml(wrapper.html())).toMatchSnapshot();
  });
});
