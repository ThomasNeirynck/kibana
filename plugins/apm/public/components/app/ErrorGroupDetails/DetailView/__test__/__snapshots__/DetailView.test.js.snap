// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailView should render empty state 1`] = `
"
null
"
`;

exports[`DetailView should render with data 1`] = `
"
<div class=\\"sc-brqgnP gzTAIt\\">
  <div class=\\"sc-cMljjf kRTmxt\\">
    <h2 class=\\"sc-jAaTju giZDJt sc-htoDjs kEDvNn\\">
      Error occurrence
    </h2>
    <a href=\\"/app/kibana#/discover?_g=&amp;_a=(interval:auto,query:(language:lucene,query:'context.service.name:opbeans-node AND error.grouping_key:c00e245c2fbebaf178fc31eeb2bb0250'),sort:('@timestamp':desc))\\"
       class=\\"kuiLink \\"
    >
      <button class=\\"kuiButton kuiButton--secondary kuiButton--iconText\\">
        <span class=\\"kuiButton__inner\\">
          <span aria-hidden=\\"true\\"
                class=\\"kuiButton__icon kuiIcon fa-compass\\"
          >
          </span>
          <span>
            View 18 occurences in Discover
          </span>
        </span>
      </button>
    </a>
  </div>
  <div class=\\"sc-bdVaJa fcOFyp\\">
    <div class=\\"sc-bwzfXH Evm\\">
      <div class=\\"sc-bxivhb cZfMoG\\">
        Timestamp
      </div>
      <div class=\\"sc-ifAKCX LYQmD\\">
        1515508740
        <span class=\\"sc-EHOje fxIYbA\\">
          (1515508740)
        </span>
      </div>
    </div>
    <div class=\\"sc-htpNat iGoPZi\\">
      <div class=\\"sc-bxivhb cZfMoG\\">
        URL
      </div>
      <span class=\\"sc-bZQynM gGoFAs\\"
            title=\\"N/A\\"
      >
        N/A
      </span>
    </div>
    <div class=\\"sc-bwzfXH Evm\\">
      <div class=\\"sc-bxivhb cZfMoG\\">
        Request method
      </div>
      <div class=\\"sc-ifAKCX LYQmD\\">
        GET
      </div>
    </div>
    <div class=\\"sc-bwzfXH Evm\\">
      <div class=\\"sc-bxivhb cZfMoG\\">
        Handled
      </div>
      <div class=\\"sc-ifAKCX LYQmD\\">
        N/A
      </div>
    </div>
    <div class=\\"sc-bwzfXH Evm\\">
      <div class=\\"sc-bxivhb cZfMoG\\">
        User ID
      </div>
      <div class=\\"sc-ifAKCX LYQmD\\">
        N/A
      </div>
    </div>
  </div>
  <div class=\\"sc-jDwBTQ krGBWu\\">
    <div class=\\"sc-iwsKbI xfglV\\">
      Log stacktrace
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      Exception stacktrace
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      Request
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      Response
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      System
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      Service
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      Process
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      User
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      Tags
    </div>
    <div class=\\"sc-iwsKbI eQURAw\\">
      Custom
    </div>
  </div>
  <div class=\\"sc-gPEVay dLfQVr\\">
    <div class=\\"kuiTableInfo sc-hSdWYo fqcxgH\\">
      No stacktrace available.
      <div class=\\"sc-eHgmQL kFyYkl\\">
        <span>
          Oops! You should try another time range. If that's no good, there's always the
          <a href=\\"https://www.elastic.co/guide/en/apm/get-started/6.2/index.html\\">
            documentation
          </a>
          .
        </span>
      </div>
    </div>
  </div>
</div>
"
`;
